File: stockvision-frontend/scripts/check-no-auth.mjs
/**
 * Regression guard: assert the frontend's authentication code has not returned.
 *
 * Checks IMPORT SPECIFIERS and identifier usage rather than raw text, because the
 * source deliberately documents the removed subsystem in comments — a naive text
 * search matches those and fails the build for describing history accurately.
 */
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

const FORBIDDEN_IMPORTS = [
  "@/store/auth-store",
  "@/hooks/use-auth",
  "@/lib/auto-auth",
  "@/components/layout/auth-guard",
];
const FORBIDDEN_IDENTIFIERS = ["useAuthStore", "useAuth", "AuthGuard", "ensureAuthenticated"];

function walk(dir) {
  return readdirSync(dir).flatMap((entry) => {
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) return walk(full);
    return /\.(ts|tsx)$/.test(entry) ? [full] : [];
  });
}

const findings = [];
for (const file of walk("src")) {
  // Strip comments so prose about the removal is not matched.
  const code = readFileSync(file, "utf8")
    .replace(/\/\*[\s\S]*?\*\//g, "")
    .replace(/^\s*\/\/.*$/gm, "");

  code.split("\n").forEach((line, index) => {
    for (const spec of FORBIDDEN_IMPORTS) {
      if (line.includes(`"${spec}"`) || line.includes(`'${spec}'`)) {
        findings.push(`${file}:${index + 1}: imports ${spec}`);
      }
    }
    for (const name of FORBIDDEN_IDENTIFIERS) {
      if (new RegExp(`\\b${name}\\s*[(<]`).test(line)) {
        findings.push(`${file}:${index + 1}: uses ${name}`);
      }
    }
  });
}

if (findings.length) {
  console.error("Authentication code found — it was removed in v2.0:\n");
  findings.forEach((f) => console.error(`  ${f}`));
  process.exit(1);
}
console.log("Clean: no authentication code in the frontend.");

File: stockvision-backend/scripts/check_no_auth.py
"""
Regression guard: assert the authentication subsystem has not returned.

Removing auth was a deliberate, wide-reaching change (see docs/CHANGELOG.md), and
the easiest way for it to creep back is one innocuous-looking import in a new
endpoint. This runs in CI.

Uses the AST rather than `grep`, because the source deliberately DISCUSSES the
removed subsystem in its CHANGE LOG docstrings — a naive text search matches those
comments and fails the build for describing history accurately. Parsing means we
check what the code *does*, not what it talks about.
"""
import ast
from pathlib import Path

FORBIDDEN_MODULES = {"jose", "passlib", "bcrypt", "app.core.security", "app.services.auth_service"}
FORBIDDEN_NAMES = {
    "OAuth2PasswordBearer", "OAuth2PasswordRequestForm",
    "get_current_user", "require_roles", "UserRole", "AuthService", "UserRepository",
}

APP_ROOT = Path(__file__).resolve().parent.parent / "app"


def module_is_forbidden(name: str | None) -> bool:
    if not name:
        return False
    return name in FORBIDDEN_MODULES or name.split(".")[0] in FORBIDDEN_MODULES


def scan(path: Path) -> list[str]:
    """Return human-readable findings for one file."""
    findings: list[str] = []
    tree = ast.parse(path.read_text(), filename=str(path))

    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                if module_is_forbidden(alias.name):
                    findings.append(f"{path}:{node.lineno}: imports `{alias.name}`")

        elif isinstance(node, ast.ImportFrom):
            if module_is_forbidden(node.module):
                findings.append(f"{path}:{node.lineno}: imports from `{node.module}`")
            for alias in node.names:
                if alias.name in FORBIDDEN_NAMES:
                    findings.append(f"{path}:{node.lineno}: imports `{alias.name}`")

        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef)):
            if node.name in FORBIDDEN_NAMES:
                findings.append(f"{path}:{node.lineno}: defines `{node.name}`")

        elif isinstance(node, ast.Call):
            name = getattr(node.func, "id", None) or getattr(node.func, "attr", None)
            if name in FORBIDDEN_NAMES:
                findings.append(f"{path}:{node.lineno}: calls `{name}`")

    return findings


def main() -> int:
    findings: list[str] = []
    for path in sorted(APP_ROOT.rglob("*.py")):
        findings.extend(scan(path))

    if findings:
        print("Authentication code found — it was removed in v2.0:\n")
        for finding in findings:
            print(f"  {finding}")
        return 1

    print(f"Clean: no authentication code across {sum(1 for _ in APP_ROOT.rglob('*.py'))} modules.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
